game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/RemoveUnprotectedUnits.java

```diff
@@ -0,0 +1,173 @@
+      if (!enemyUnits.isEmpty()) {
+        final Change change =
+            ChangeFactory.markNoMovementChange(
+                CollectionUtils.getMatches(enemyUnits, Matches.unitIsSea()));
+        bridge.addChange(change);
+        battleActions.remove(
+            alliedUnits, bridge, battleState.getBattleSite(), side == BattleState.Side.DEFENSE);
+      }
+    }
+  }
+
+  private boolean attackerHasRetreat(final BattleState.Side side) {
+    return side == BattleState.Side.OFFENSE
+        && (!battleState.getAttackerRetreatTerritories().isEmpty()
+            || battleState.getAttackingUnits().stream().anyMatch(Matches.unitIsAir()));
+  }
+
+  private List<Unit> getAlliedTransports(final GamePlayer player) {
+    final Predicate<Unit> matchAllied =
+        Matches.unitIsTransport()
+            .and(Matches.unitIsNotCombatTransport())
+            .and(Matches.isUnitAllied(player, battleState.getGameData()))
+            .and(Matches.unitIsSea());
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), matchAllied);
+  }
+
+  private Collection<Unit> getAlliedUnits(final GamePlayer player) {
+    final Predicate<Unit> alliedUnitsMatch =
+        Matches.isUnitAllied(player, battleState.getGameData())
+            .and(Matches.unitIsNotLand())
+            .and(Matches.unitIsSubmerged().negate());
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), alliedUnitsMatch);
+  }
+
+  private Collection<Unit> getEnemyUnitsThatCanFire(final GamePlayer player) {
+    final Predicate<Unit> enemyUnitsMatch =
+        Matches.unitIsNotLand()
+            .and(Matches.enemyUnit(player, battleState.getGameData()))
+            .and(Matches.unitIsSubmerged().negate())
+            .and(Matches.unitCanAttack(player));
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), enemyUnitsMatch);
+  }
+
+  private void checkUnprotectedUnits(final IDelegateBridge bridge, final BattleState.Side side) {
+    // if we are the attacker, we can retreat instead of dying
+    if (attackerHasRetreat(side)) {
+      return;
+    }
+    if (battleState
+        .getUnits(EnumSet.of(BattleState.Side.OFFENSE, BattleState.Side.DEFENSE))
+        .isEmpty()) {
+      return;
+    }
+    final Collection<Unit> unprotectedUnits = getUnprotectedUnits(side);
+    final boolean hasUnitsThatCanRollLeft = areFightingOrSupportingUnitsLeft(side);
+    final boolean enemyHasUnitsThatCanRollLeft =
+        areFightingOrSupportingUnitsLeft(side.getOpposite());
+    if (!hasUnitsThatCanRollLeft && enemyHasUnitsThatCanRollLeft) {
+      battleActions.remove(
+          unprotectedUnits, bridge, battleState.getBattleSite(), side == BattleState.Side.DEFENSE);
+    }
+  }
+
+  private boolean areFightingOrSupportingUnitsLeft(final BattleState.Side side) {
+    final boolean hasUnitsThatCanRollLeft;
+    hasUnitsThatCanRollLeft =
+        battleState.getUnits(EnumSet.of(side)).stream()
+            .anyMatch(
```

Comment:
nit, deeply nested functional chain. Consider decomposing this a bit with helper methods. EG:

```
return battleState.getUnits(EnumSet.of(side)).stream()
  .anyMatch(unitIsActive(BattleState.side.OFFENSE));
```

I believe you can also get rid of the intermediate variable as well and just return the result directly as well.